File: src/js/core/src/Style/Canvas.ts
import type { StyleDefinitionCanvas } from '../StyleDefinition.js';
import { Element } from './Element.js';

/**
 * Read-only view over a style definition's `canvas` block, exposing the
 * drawing area dimensions and the top-level element list.
 */
export class Canvas {
  #data: StyleDefinitionCanvas;
  #elements?: readonly Element[];

  constructor(data: StyleDefinitionCanvas) {
    this.#data = data;
  }

  /**
   * Returns the canvas width — the `width` value of the SVG `viewBox`.
   */
  width(): number {
    return this.#data.width;
  }

  /**
   * Returns the canvas height — the `height` value of the SVG `viewBox`.
   */
  height(): number {
    return this.#data.height;
  }

  /**
   * Returns the top-level elements rendered onto the canvas, lazily wrapped
   * as {@link Element} instances on first access.
   */
  elements(): readonly Element[] {
    this.#elements ??= this.#data.elements.map((el) => new Element(el));

    return this.#elements;
  }
}

File: src/rust/core/src/style/canvas.rs
//! Read-only view over a style definition's `canvas` block.

use serde::Deserialize;

use super::element::Element;

#[derive(Deserialize)]
pub(crate) struct Canvas {
    width: f64,
    height: f64,
    elements: Vec<Element>,
}

impl Canvas {
    pub(crate) fn width(&self) -> f64 {
        self.width
    }

    pub(crate) fn height(&self) -> f64 {
        self.height
    }

    pub(crate) fn elements(&self) -> &[Element] {
        &self.elements
    }
}
